### ISSUE TITLE
Client Modify Binary Property  

### ISSUE THREAD

**Issue by DaleCox** [mcavage/node-ldapjs#333]
Thanks for creating this library it’s really useful. 

I’m trying to replicate the .NET code for updating the thumbnail photo with node. I’m made pretty good progress but the octet string from the node buffer object is different than the octet string the .NET  byte array produces. 

Can you provide any guidance on how to read the image file so that the stored octet strings are the same? 

.NET code is here http://blogs.msdn.com/b/rakhiguha/archive/2015/09/28/active-directory-user-thumbnail-photo-update-using-service-account.aspx

Here is the node code I am running

var ChangePhoto = function(client, domain, img, idsid){
	var deferred = Q.defer();
	
    //console.log('img ', img);//debug
	var change = new LDAP.Change({
		operation: 'replace',
		modification: {
			'thumbnailPhoto': img
		}
	});
	
	var dn = 'cn=foo, o=example' ;
	console.log('dn ', dn);
	
	client.modify(dn, change, function(err) {
		if(err){
			console.error(err);
			deferred.reject(err);
		}else{
			console.log("success");
			deferred.resolve(true);
		}
	});
	
	return deferred.promise;
};

**Comment by pfmooney** [comment]
What do you mean by the octet string differing.  Is `img` a buffer object which you've read directly from the file in question.  By what means are you verifying that the updated object in LDAP is or is not correct after the modification.

(As a side note, please wrap code segments in \`\`\` code \`\`\` tags)

**Comment by DaleCox** [comment]
The img is a node buffer object unencoded from fs.readFileSync. 

I am using ADExplorer to pull the octet string from thumbnailPhoto attribute which is how I compared the strings after running each algorithm. 

It seems that the node buffer object and the .NET byte arrays are producing two different octet strings. 

I am assuming that each library is simply passing the octet string through which is why I was asking if I should encode the node buffer object in a certain way to make it match the .NET value.

Code sample markup updated, thanks for the pointer.

**Comment by pfmooney** [comment]
So I assume that the byte string present in LDAP after the .NET program runs is equal to the file present on disk?  If you believe the data from ldapjs is being somehow altered as part of the transmission process, I would suggest collecting some trace-level bunyan logs.

**Comment by sinistersig** [comment]
I think this may be a bug in the library, I had this same problem this morning.

https://github.com/mcavage/node-ldapjs/blob/540a3a1f5c2f23ade95142fa2f43fb87db585b48/lib/change.js#L87

`_attr.addValue(val[k].toString());` 
converts buffers to strings which causes an issue when it is added as an attribute

https://github.com/mcavage/node-ldapjs/blob/540a3a1f5c2f23ade95142fa2f43fb87db585b48/lib/attribute.js#L71

`if (Buffer.isBuffer(val))` will always be false

I added a check to ensure that it doesn't call `.toString()` if the value is a buffer
```
var currVal = val[k];
var attrVal = Buffer.isBuffer(currVal) ? currVal : currVal.toString();
_attr.addValue(attrVal);
```

This question, although doesn't use ldapjs, has more details on the encoding issue:
http://stackoverflow.com/questions/19569986/node-js-uploading-output-of-imagemagick-to-aws-s3

I am working on PR let me know what you guys think.

**Comment by sinistersig** [comment]
Also, I tried adding the `;binary` suffix to the type and that didn't work.

**Comment by sinistersig** [comment]
Is this still an issue?

**Comment by simonmilz** [comment]
@sinistersig yep. I'm still not able to properly save the image or convert it into a proper format node could handle.

**Comment by jsumners** [comment]
PR #556 has been merged into the `next` branch to address this.

**Comment by jsumners** [comment]
⚠️ This issue has been locked due to age. If you have encountered a recent
problem that seems to be covered by this issue, please open a new issue.

Please include a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)
when opening a new issue.
